### PR TITLE
Ported to Windows.

### PR DESCRIPTION
**Pull request by JohnEricson**
I would like to merge my changes in this module. I have ported the module so it now also works on Windows and used it successfully in a mixed Windows/Linux environment for several months. In order to minimize the changes required in order to make it run on Windows I've relied on Cygwin for providing the Unix commands needed for the many execs. Where packages is needed I've used Chocolatey which is a package management tool for Windows.

I'm relying on the changes in this module for making the Puppet-Pentaho (https://github.com/JohnEricson/puppet-pentaho) module working on both Linux and Windows.

### CODE REVIEW

**Comment by HelenCampbell** [comment]
Hi,
This will definitely need to be reviewed by the community due to the large number of changes. Thank you for the great effort you put into this. 
One thing to note is that you will need to run puppet lint against this, like so:
metadata-json-lint metadata.json
What are the Joe comments for ?
How well do the acceptance tests run in windows ?

File: manifests/client/install.pp [hunk 1]
```diff
@@ -2,13 +2,14 @@
...
-
+		if $::osfamily == 'Windows' {  
+			include mysql::server::install
+		} else {
+			package { 'mysql_client':
+				ensure =>          $mysql::client::package_ensure,
+        install_options => $mysql::client::install_options,
+				name   =>          $mysql::client::package_name,
+			}  
+		}
```
Comment (igalic) [inline manifests/client/install.pp]: all of your indentation is broken here… please refer to the [styleguide](https://docs.puppetlabs.com/guides/style_guide.html)

File: manifests/client/install.pp [hunk 2]
```diff
@@ -2,13 +2,14 @@
...
   if $mysql::client::package_manage {
-
-    package { 'mysql_client':
-      ensure          => $mysql::client::package_ensure,
-      install_options => $mysql::client::install_options,
-      name            => $mysql::client::package_name,
-    }
-
+		if $::osfamily == 'Windows' {  
+			include mysql::server::install
```
Comment (igalic) [inline manifests/client/install.pp]: why is it necessary to install mysql::server on a client anyway??

File: manifests/db.pp [hunk 3]
```diff
@@ -12,6 +12,20 @@
...
   $import_timeout = 300,
 ) {
+  
+  # JOE
+  if $::osfamily == 'Windows' {
+    Exec { provider => 'cygwin' }
+    # Use native mysql client command that is shipped with the Chococaley package. Alternatively 
+    # we could install the Cygwin package, but we would need to make sure we used same versions for best practise.
+    # TODO: Earlier I didn't need to hardcode the path.
+    $mysql_command = 'C:/tools/mysql/current/bin/mysql'
```
Comment (igalic) [inline manifests/db.pp]: why is the provider cygwin and not powershell?

**Comment by igalic** [comment]
@JohnEricson thank you very much for this patch, i added a couple of comments about the code.

general questions: Is chocolatey the canonical package manager for windows?
/cc @cyberious

**Comment by JohnEricson** [comment]
Thank you @HelenCampbell  and @igalic. I will get back to you once I've looked into these issues.

Regarding Chocolatey. No Chcocolatey (https://chocolatey.org/) is an open source initiative to provide an apt-get/yum equivalent for Windows. Officially I don't yet think Windows has a package format that provides a central repository. The closest to a package format is probably MSI-packages but no central repository exists yet for this which provides for functionality like "apt-get install vim", or as in Chocolatey you write "cinst vim" in a cmd.exe or PowerShell window.

I find Chocolatey to be the most mature solution today, and Microsoft is actually supporting it in their OneGet (https://github.com/OneGet/oneget) solution they're currently developing. OneGet is open source and unlike typical Linux solutions OneGet is a unified interface to different package management systems, such as MSI, NuGet, Chocolatey and more. Currently Microsoft has implemented support for Chocolatey in OneGet so you easily can install from cental repository.
I expect OneGet to be available in future versions of Windows 10 and equivalent Windows Server OS.

File: metadata.json [hunk 4]
```diff
@@ -84,6 +84,8 @@
   "description": "Mysql module",
   "dependencies": [
     {"name":"puppetlabs/stdlib","version_requirement":">= 3.2.0 < 5.0.0"},
-    {"name":"nanliu/staging","version_requirement":">= 1.0.1 < 2.0.0"}
+    {"name":"nanliu/staging","version_requirement":">= 1.0.1 < 2.0.0"},
+    {"name":"johnericson/cygwin_exec","version_requirement":">=0.1.0"},
+    {"name":"rismoney/chocolatey","version_requirement":">=0.4.0"}
   ]
```
Comment (igalic) [inline metadata.json]: these dependencies attempt necessary as coire requirement for each platform, so espoused only document them in the README.

**Comment by tphoney** [comment]
Hi @JohnEricson , thanks for the work you have put into this pull request. However due to inactivity and failing unit tests / needs rebased this pull request is being closed. If you want to re-submit this PR with the necessary changes, we will gladly look at the new pull request.

**Comment by cyberious** [comment]
Love the effort if @johnericson can change to Powershell and cleanup

**Comment by igalic** [comment]
+1

**Comment by hunner** [comment]
Yes, that would be wonderful if someone could pick this up again! :D

**Pull request closed without merging**